### PR TITLE
PackageModel: relative paths for cross-compilation destinations

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Depends on #5871.

This introduces support for relative paths in v2 scheme of `destination.json` files. The path is then normalized relative to the location of `destination.json` file. 

### Motivation:

Currently, `destination.json` files only support absolute paths, which makes it hard to move cross-compilation sysroots between directories and copying them to different machines.

### Modifications:

`struct DestinationInfo` is renamed to `DestinationInfoV1`. New `struct DestinationInfoV2` is added to handle the new version that supports relative paths. This is also used as an opportunity to make the schema more consistent with manifests for artifact bundles, in which properties use `camelCase` instead of `dash-case` naming. Also new in schema version 2 are `linkerFlags` and renaming of `extraCPPFlags` to `extraCXXFlags`, to follow up on #5827 and #5837.

Added tests to verify that both schema versions are parsed correctly and that relative paths in V2 are correctly expanded to absolute paths.

### Result:

New version of `destination.json` files can be supplied as a `--destination` option passed to `swift build`.

As a result, it's to move cross-compilation sysroots and their corresponding `destination.json` files between different directories on the same filesystem, or even between different machines.

### CODE REVIEW

File: Sources/PackageModel/Destination.swift [hunk 1]
```diff
@@ -212,27 +213,51 @@ extension Destination {
...
-        )
+    }
+}
+
+fileprivate extension FilePath {
+    func absolutePath(relativeTo directory: AbsolutePath) -> AbsolutePath {
+        if isRelative {
+            return directory.appending(RelativePath(string))
+        } else {
+            return AbsolutePath(string)
```
Comment (neonichu) [inline Sources/PackageModel/Destination.swift:251]: This should use `AbsolutePath(validating:)`

**Review by stevapple — COMMENTED** [review]
Is there any reason we cannot add it in v1 scheme? Supporting relative paths looks like a pure addition.

**Comment by MaxDesiatov** [comment]
This PR is still a draft, there highly likely will be incompatible changes in future commits.

File: Sources/PackageModel/Destination.swift [hunk 2]
```diff
@@ -232,28 +233,56 @@ extension Destination {
     public init(fromFile path: AbsolutePath, fileSystem: FileSystem) throws {
         let decoder = JSONDecoder.makeWithDefaults()
         let version = try decoder.decode(path: path, fileSystem: fileSystem, as: VersionInfo.self)
+        
         // Check schema version.
-        guard version.version == 1 else {
+        if version.version == 1 {
```
Comment (tomerd) [inline Sources/PackageModel/Destination.swift:238]: use switch?

File: Sources/PackageModel/Destination.swift [hunk 3]
```diff
@@ -232,28 +233,56 @@ extension Destination {
...
+    /// Converts `FilePath` to `AbsolutePath`, in relation to other `AbsolutePath` if possible.
+    /// - Parameter directory: an absolute path to a directory used to calculate a resulting `AbsolutePath`.
+    /// - Returns: trivially converts `self` to `AbsolutePath` as a string if `self` is an absolute path.
+    /// Otherwise, if `self` is a relative path, returns `self` appended to the `directory` argument.
+    func absolutePath(relativeTo directory: AbsolutePath) throws -> AbsolutePath {
+        if isRelative {
+            return directory.appending(RelativePath(string))
+        } else {
+            return try AbsolutePath(validating: string)
+        }
```
Comment (tomerd) [inline Sources/PackageModel/Destination.swift:277]: iirc the absolute path constructor already takes care of it? @neonichu ?

Comment (neonichu) [reply Sources/PackageModel/Destination.swift:277]: Yah, `init(validating:, relativeTo:)` should do the same thing, I believe

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Review by neonichu — APPROVED** [review]

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@MaxDesiatov is this ready to be merged?

**Comment by MaxDesiatov** [comment]
I'm waiting for https://github.com/apple/swift-package-manager/pull/5876 to be merged, otherwise this PR will cause conflicts in that one. Unless @neonichu is fine with that? IDK how much more #5876 is going to take though.

**Comment by MaxDesiatov** [comment]
After double-checking, it seems both PRs don't overlap on specific lines, although `Destination.swift` is modified in both. I'll go ahead and merge this one.

**Pull request merged**